File: src/gestorAplicacion/Registro.java
package gestorAplicacion;

import baseDatos.Deserializador;
import java.util.ArrayList;
import java.io.Serializable;

public abstract class Registro implements Serializable{
    private static final long serialVersionUID = 1L;
    public static short cantidadUsuariosExistentes;
    private String nombre;
    private String correo;
    private String nombreUsuario;
    private String clave;
    private String documentoIdentificacion;
    private static ArrayList<Estudiante> estudiantes = new ArrayList<Estudiante>();
    private static ArrayList<Estudiante> estudiantesMatriculados = new ArrayList<Estudiante>();
    private static ArrayList<Curso> cursos = new ArrayList<Curso>();
    private static ArrayList<Admin> admins = new ArrayList<Admin>();
    private static ArrayList<Profesor> profesores = new ArrayList<Profesor>();
    private static ArrayList<EstimuloEstudiante> estimulosEstudiantes = new ArrayList<EstimuloEstudiante>();
    private static ArrayList<EstimuloProfesor> estimulosProfesores = new ArrayList<EstimuloProfesor>();
    private static ArrayList<Estudiante> estudiantesConCita = new ArrayList<Estudiante>();

    public Registro(String nombre, String correo, String nombreUsuario, String clave, String documentoIdentificacion) {
        this.nombre = nombre;
        this.correo = correo;
        this.nombreUsuario = nombreUsuario;
        this.clave = clave;
        this.documentoIdentificacion = documentoIdentificacion;
    }

    public static ArrayList<Estudiante> getEstudiantesConCita() {
        return estudiantesConCita;
    }

    public static void setEstudiantesConCita(ArrayList<Estudiante> estudiantesConCita) {
        Registro.estudiantesConCita = estudiantesConCita;
    }

    public static short getCantidadUsuariosExistentes() {
        return cantidadUsuariosExistentes;
    }

    public static void setCantidadUsuariosExistentes(short cantidadUsuariosExistentes) {
        Registro.cantidadUsuariosExistentes = cantidadUsuariosExistentes;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public String getCorreo() {
        return correo;
    }

    public void setCorreo(String correo) {
        this.correo = correo;
    }

    public String getNombreUsuario() {
        return nombreUsuario;
    }

    public void setNombreUsuario(String nombreUsuario) {
        this.nombreUsuario = nombreUsuario;
    }

    public String getClave() {
        return clave;
    }

    public void setClave(String clave) {
        this.clave = clave;
    }

    public String getDocumentoIdentificacion() {
        return documentoIdentificacion;
    }

    public void setDocumentoIdentificacion(String documentoIdentificacion) {
        this.documentoIdentificacion = documentoIdentificacion;
    }

    
    public static ArrayList<Estudiante> getEstudiantes() {
        return estudiantes;
    }

    public static void setEstudiantes(ArrayList<Estudiante> estudiantes) {
        Registro.estudiantes = estudiantes;
    }
    
    public static void agregarEstudiante(Estudiante estudiante){
        Registro.estudiantes.add(estudiante);
    }

    
    public static ArrayList<Estudiante> getEstudiantesMatriculados() {
        return estudiantesMatriculados;
    }

    public static void setEstudiantesMatriculados(ArrayList<Estudiante> estudiantesMatriculados) {
        Registro.estudiantesMatriculados = estudiantesMatriculados;
    }
    
    public static void agregarEstudianteMatriculado(Estudiante estudiante){
        Registro.estudiantesMatriculados.add(estudiante);
    }
    
    
    public static ArrayList<Curso> getCursos() {
        return cursos;
    }

    public static void setCursos(ArrayList<Curso> cursos) {
        Registro.cursos = cursos;
    }
    
    public static void agregarCurso(Curso curso) {
        boolean comp = true;
        for(Curso c : cursos){
            if(c.getNombre().equals(curso.getNombre())){
                comp = false;
                break;
            }  
        }
        if(comp == true){
            cursos.add(curso);
        }
    }
    
    public static void eliminarCurso(Curso curso) {
        cursos.remove(curso);
    }

    public static void eliminarCurso(int i) {
        cursos.remove(i);
    }
    
    public static ArrayList<Admin> getAdmins() {
        return admins;
    }

    public static void setAdmins(ArrayList<Admin> admins) {
        Registro.admins = admins;
    }
    
    public static void agregarAdmin(Admin admin){
        Registro.admins.add(admin);
    }

    
    public static ArrayList<Profesor> getProfesores() {
        return profesores;
    }

    public static void setProfesores(ArrayList<Profesor> profesores) {
        Registro.profesores = profesores;
    }
    
    public static void agregarProfesor(Profesor profesor){
        Registro.profesores.add(profesor);
    }

    public static ArrayList<EstimuloEstudiante> getEstimulosEstudiantes() {
        return estimulosEstudiantes;
    }

    public static void setEstimulosEstudiantes(ArrayList<EstimuloEstudiante> estimulos) {
        Registro.estimulosEstudiantes = estimulos;
    }
    
    public static void agregarEstimulosEstudiantes(EstimuloEstudiante estimulos){
        Registro.estimulosEstudiantes.add(estimulos);
    }

    public static ArrayList<EstimuloProfesor> getEstimulosProfesores() {
        return estimulosProfesores;
    }

    public static void setEstimulosProfesores(ArrayList<EstimuloProfesor> estimulos) {
        Registro.estimulosProfesores = estimulos;
    }
    
    public static void agregarEstimulosProfesores(EstimuloProfesor estimulos){
        Registro.estimulosProfesores.add(estimulos);
    }
    
    public abstract void buscarCursos();
}

File: PYTHON/src/gestorAplicacion/clasesDeUsuario/Registro.py
from abc import ABC, abstractclassmethod
from gestorAplicacion.clasesDeCurso.Curso import Curso
class Registro(ABC):
    _cantidadUsuariosExistentes = 0
    _estudiantes = []
    _estudiantesMatriculados = []
    _cursos = []
    _admins = []
    _profesores = []
    _estimulosEstudiantes = []
    _estimulosProfesores = []
    _estudiantesConCita = []

    def __init__(self, nombre, correo, nombreUsuario, clave, documentoIdentificacion):
        self._nombre = nombre
        self._correo = correo
        self._nombreUsuario = nombreUsuario
        self._clave = clave
        self._documentoIdentificacion = documentoIdentificacion
    
    @classmethod
    def getEstudiantesConCita(cls):
        return Registro._estudiantesConCita
    
    @classmethod
    def setEstudiantesConCita(cls, estudiantesConCita):
        Registro._estudiantesConCita = estudiantesConCita

    @classmethod
    def getCantidadUsuariosExistentes(cls):
        return Registro._cantidadUsuariosExistentes

    @classmethod
    def setCantidadUsuariosExistentes(cls, cantidadUsuariosExistentes):
        Registro._cantidadUsuariosExistentes = cantidadUsuariosExistentes
    
    @classmethod
    def getAdmins(cls): 
        return Registro._admins
    
    @classmethod
    def setAdmins(cls, admins):
        Registro._admins = admins
    
    @classmethod
    def agregarAdmin(cls, admin):
        Registro._admins.append(admin)
    
    @classmethod
    def getProfesores(cls):
        return Registro._profesores
    
    @classmethod
    def setProfesores(cls, profesores):
        Registro._profesores = profesores
    
    @classmethod
    def agregarProfesor(cls, profesor):
        Registro._profesores.append(profesor)
    
    @classmethod
    def getEstimulosEstudiantes(cls):
        return Registro._estimulosEstudiantes
    
    @classmethod
    def setEstimulosEstudiantes(cls, estimulos):
        Registro._estimulosEstudiantes = estimulos
    
    @classmethod
    def agregarEstimulosEstudiantes(cls, estimulos):
        Registro._estimulosEstudiantes.append(estimulos)
    
    @classmethod
    def getEstimulosProfesores(cls):
        return Registro._estimulosProfesores
    
    @classmethod
    def setEstimulosProfesores(cls, estimulos):
        Registro._estimulosProfesores = estimulos
    
    @classmethod
    def agregarEstimulosProfesores(cls, estimulos):
        Registro._estimulosProfesores.append(estimulos)
    
    @classmethod
    def getEstudiantes(cls):
        return Registro._estudiantes
    
    @classmethod
    def setEstudiantes(cls, estudiantes):
        Registro._estudiantes = estudiantes

    @classmethod
    def agregarEstudiante(cls, estudiante):
        Registro._estudiantes.append(estudiante)
    
    @classmethod
    def getEstudiantesMatriculados(cls):
        return Registro._estudiantesMatriculados
    
    @classmethod
    def setEstudiantesMatriculados(cls, estudiantesMatriculados):
        Registro._estudiantesMatriculados = estudiantesMatriculados
    
    @classmethod
    def agregarEstudianteMatriculado(cls, estudiante):
        Registro._estudiantesMatriculados.append(estudiante)
    
    @classmethod
    def getCursos(cls) -> list[Curso] : 
        return Registro._cursos
    
    @classmethod
    def setCursos(cls, cursos):
        Registro._cursos = cursos
    
    @classmethod
    def agregarCurso(cls, curso):
        comp = True
        for c in Registro._cursos:
            if c.getNombre() == curso.getNombre():
                comp = False
                break
        if comp == True:
            Registro._cursos.append(curso)
    
    @classmethod
    def eliminarCurso(cls, curso):
        Registro._cursos.remove(curso)
    
    @classmethod
    def eliminarCurso(cls, i): #En desuso
        Registro._cursos.remove(i)
    
    def getNombre(self):
        return self._nombre
    
    def setNombre(self, nombre):
        self._nombre = nombre
    

    def getCorreo(self):
        return self._correo
    
    def setCorreo(self, correo):
        self._correo = correo
    
    def getNombreUsuario(self):
        return self._nombreUsuario
    
    def setNombreUsuario(self, nombreUsuario): 
        self._nombreUsuario = nombreUsuario
    
    def getClave(self):
        return self._clave
    
    def setClave(self, clave): 
        self._clave = clave
    
    def getDocumentoIdentificacion(self): 
        return self._documentoIdentificacion
    
    def setDocumentoIdentificacion(self, documentoIdentificacion): 
        self._documentoIdentificacion = documentoIdentificacion
    
    @abstractclassmethod
    def buscarCursos(self):
        pass
